Content.Server/VendingMachines/VendingMachineComponent.cs

```diff
@@ -198,10 +249,56 @@ private void TryEject(string id, EntityUid? sender)
                 {
                     Owner.PopupMessageEveryone(Loc.GetString("vending-machine-component-try-eject-access-denied"));
                     Deny();
-                    return;
+                    return false;
                 }
             }
-            TryEject(id);
+            return true;
+        }
+
+        private void AuthorizedVend(string id, EntityUid? sender)
+        {
+            if (IsAuthorized(sender))
+            {
+                TryDispense(id);
+            }
+            return;
+        }
+
+        async Task<bool> IInteractUsing.InteractUsing(InteractUsingEventArgs eventArgs)
+        {
+            if (_storage == null || Whitelist == null) {
+                return false;
+            }
+            if (!IsAuthorized(eventArgs.User))
+            {
+                return false;
+            }
+            if (_entMan.GetComponent<HandsComponent>(eventArgs.User).GetActiveHand?.Owner is not {Valid: true} itemEntity)
+            {
+                eventArgs.User.PopupMessage(Loc.GetString("vending-machine-component-interact-using-no-active-hand"));
+                return false;
+            }
+            if (!Whitelist.IsValid(itemEntity))
+            {
+                return false;
+            }
+            if (!_entMan.TryGetComponent(itemEntity, typeof(ItemComponent), out var item))
+            {
+                return false;
+            }
+            var metaData = _entMan.GetComponent<MetaDataComponent>(item.Owner);
+            if (metaData.EntityPrototype == null)
+            {
+                return false;
+            }
```

Comment:
Functionality should never be based on a prototype's ID. It's only to be used as an identifier for spawning and nothing more. This is why entity whitelist doesn't even support it.